File: photon-lib/py/photonlibpy/estimation/targetModel.py
import math
from typing import List, Self

from wpimath.geometry import Pose3d, Rotation2d, Rotation3d, Translation3d
from wpimath.units import meters

from . import RotTrlTransform3d


class TargetModel:
    """Describes the 3d model of a target."""

    def __init__(self):
        """Default constructor for initialising internal class members. DO NOT USE THIS!!! USE THE createPlanar,
        createCuboid, createSpheroid or create Arbitrary
        """
        self.vertices: List[Translation3d] = []
        self.isPlanar = False
        self.isSpherical = False

    @classmethod
    def createPlanar(cls, width: meters, height: meters) -> Self:
        """Creates a rectangular, planar target model given the width and height. The model has four
        vertices:

        - Point 0: [0, -width/2, -height/2]
        - Point 1: [0, width/2, -height/2]
        - Point 2: [0, width/2, height/2]
        - Point 3: [0, -width/2, height/2]
        """

        tm = cls()

        tm.isPlanar = True
        tm.isSpherical = False
        tm.vertices = [
            Translation3d(0.0, -width / 2.0, -height / 2.0),
            Translation3d(0.0, width / 2.0, -height / 2.0),
            Translation3d(0.0, width / 2.0, height / 2.0),
            Translation3d(0.0, -width / 2.0, height / 2.0),
        ]
        return tm

    @classmethod
    def createCuboid(cls, length: meters, width: meters, height: meters) -> Self:
        """Creates a cuboid target model given the length, width, height. The model has eight vertices:

        - Point 0: [length/2, -width/2, -height/2]
        - Point 1: [length/2, width/2, -height/2]
        - Point 2: [length/2, width/2, height/2]
        - Point 3: [length/2, -width/2, height/2]
        - Point 4: [-length/2, -width/2, height/2]
        - Point 5: [-length/2, width/2, height/2]
        - Point 6: [-length/2, width/2, -height/2]
        - Point 7: [-length/2, -width/2, -height/2]
        """

        tm = cls()
        verts = [
            Translation3d(length / 2.0, -width / 2.0, -height / 2.0),
            Translation3d(length / 2.0, width / 2.0, -height / 2.0),
            Translation3d(length / 2.0, width / 2.0, height / 2.0),
            Translation3d(length / 2.0, -width / 2.0, height / 2.0),
            Translation3d(-length / 2.0, -width / 2.0, height / 2.0),
            Translation3d(-length / 2.0, width / 2.0, height / 2.0),
            Translation3d(-length / 2.0, width / 2.0, -height / 2.0),
            Translation3d(-length / 2.0, -width / 2.0, -height / 2.0),
        ]

        tm._common_construction(verts)

        return tm

    @classmethod
    def createSpheroid(cls, diameter: meters) -> Self:
        """Creates a spherical target model which has similar dimensions regardless of its rotation. This
        model has four vertices:

        - Point 0: [0, -radius, 0]
        - Point 1: [0, 0, -radius]
        - Point 2: [0, radius, 0]
        - Point 3: [0, 0, radius]

        *Q: Why these vertices?* A: This target should be oriented to the camera every frame, much
        like a sprite/decal, and these vertices represent the ellipse vertices (maxima). These vertices
        are used for drawing the image of this sphere, but do not match the corners that will be
        published by photonvision.
        """

        tm = cls()

        tm.isPlanar = False
        tm.isSpherical = True
        tm.vertices = [
            Translation3d(0.0, -diameter / 2.0, 0.0),
            Translation3d(0.0, 0.0, -diameter / 2.0),
            Translation3d(0.0, diameter / 2.0, 0.0),
            Translation3d(0.0, 0.0, diameter / 2.0),
        ]

        return tm

    @classmethod
    def createArbitrary(cls, verts: List[Translation3d]) -> Self:
        """Creates a target model from arbitrary 3d vertices. Automatically determines if the given
        vertices are planar(x == 0). More than 2 vertices must be given. If this is a planar model, the
        vertices should define a non-intersecting contour.

        :param vertices: Translations representing the vertices of this target model relative to its
                         pose.
        """

        tm = cls()
        tm._common_construction(verts)

        return tm

    def _common_construction(self, verts: List[Translation3d]) -> None:
        self.isSpherical = False
        if len(verts) <= 2:
            self.vertices = []
            self.isPlanar = False
        else:
            cornersPlaner = True
            for corner in verts:
                if abs(corner.X() < 1e-4):
                    cornersPlaner = False
            self.isPlanar = cornersPlaner

        self.vertices = verts

    def getFieldVertices(self, targetPose: Pose3d) -> List[Translation3d]:
        """This target's vertices offset from its field pose.

        Note: If this target is spherical, use {@link #getOrientedPose(Translation3d,
        Translation3d)} with this method.
        """

        basisChange = RotTrlTransform3d(targetPose.rotation(), targetPose.translation())

        retVal = []

        for vert in self.vertices:
            retVal.append(basisChange.applyTranslation(vert))

        return retVal

    @classmethod
    def getOrientedPose(cls, tgtTrl: Translation3d, cameraTrl: Translation3d):
        """Returns a Pose3d with the given target translation oriented (with its relative x-axis aligned)
        to the camera translation. This is used for spherical targets which should not have their
        projection change regardless of their own rotation.

        :param tgtTrl:    This target's translation
        :param cameraTrl: Camera's translation

        :returns: This target's pose oriented to the camera
        """

        relCam = cameraTrl - tgtTrl
        orientToCam = Rotation3d(
            0.0,
            Rotation2d(math.hypot(relCam.X(), relCam.Y()), relCam.Z()).radians(),
            Rotation2d(relCam.X(), relCam.Y()).radians(),
        )
        return Pose3d(tgtTrl, orientToCam)

    def getVertices(self) -> List[Translation3d]:
        return self.vertices

    def getIsPlanar(self) -> bool:
        return self.isPlanar

    def getIsSpherical(self) -> bool:
        return self.isSpherical

    @classmethod
    def AprilTag36h11(cls) -> Self:
        return cls.createPlanar(width=6.5 * 0.0254, height=6.5 * 0.0254)

    @classmethod
    def AprilTag16h5(cls) -> Self:
        return cls.createPlanar(width=6.0 * 0.0254, height=6.0 * 0.0254)

File: photon-targeting/src/main/java/org/photonvision/estimation/TargetModel.java
/*
 * Copyright (C) Photon Vision.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.photonvision.estimation;

import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Translation3d;
import edu.wpi.first.math.util.Units;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/** Describes the 3d model of a target. */
public class TargetModel {
    /**
     * Translations of this target's vertices relative to its pose. Rectangular and spherical targets
     * will have four vertices. See their respective constructors for more info.
     */
    public final List<Translation3d> vertices;

    public final boolean isPlanar;
    public final boolean isSpherical;

    public static final TargetModel kAprilTag16h5 =
            new TargetModel(Units.inchesToMeters(6), Units.inchesToMeters(6));
    public static final TargetModel kAprilTag36h11 =
            new TargetModel(Units.inchesToMeters(6.5), Units.inchesToMeters(6.5));

    /**
     * Creates a rectangular, planar target model given the width and height. The model has four
     * vertices:
     *
     * <ul>
     *   <li>Point 0: [0, -width/2, -height/2]
     *   <li>Point 1: [0, width/2, -height/2]
     *   <li>Point 2: [0, width/2, height/2]
     *   <li>Point 3: [0, -width/2, height/2]
     * </ul>
     */
    public TargetModel(double widthMeters, double heightMeters) {
        this.vertices =
                List.of(
                        // this order is relevant for AprilTag compatibility
                        new Translation3d(0, -widthMeters / 2.0, -heightMeters / 2.0),
                        new Translation3d(0, widthMeters / 2.0, -heightMeters / 2.0),
                        new Translation3d(0, widthMeters / 2.0, heightMeters / 2.0),
                        new Translation3d(0, -widthMeters / 2.0, heightMeters / 2.0));
        this.isPlanar = true;
        this.isSpherical = false;
    }

    /**
     * Creates a cuboid target model given the length, width, height. The model has eight vertices:
     *
     * <ul>
     *   <li>Point 0: [length/2, -width/2, -height/2]
     *   <li>Point 1: [length/2, width/2, -height/2]
     *   <li>Point 2: [length/2, width/2, height/2]
     *   <li>Point 3: [length/2, -width/2, height/2]
     *   <li>Point 4: [-length/2, -width/2, height/2]
     *   <li>Point 5: [-length/2, width/2, height/2]
     *   <li>Point 6: [-length/2, width/2, -height/2]
     *   <li>Point 7: [-length/2, -width/2, -height/2]
     * </ul>
     */
    public TargetModel(double lengthMeters, double widthMeters, double heightMeters) {
        this(
                List.of(
                        new Translation3d(lengthMeters / 2.0, -widthMeters / 2.0, -heightMeters / 2.0),
                        new Translation3d(lengthMeters / 2.0, widthMeters / 2.0, -heightMeters / 2.0),
                        new Translation3d(lengthMeters / 2.0, widthMeters / 2.0, heightMeters / 2.0),
                        new Translation3d(lengthMeters / 2.0, -widthMeters / 2.0, heightMeters / 2.0),
                        new Translation3d(-lengthMeters / 2.0, -widthMeters / 2.0, heightMeters / 2.0),
                        new Translation3d(-lengthMeters / 2.0, widthMeters / 2.0, heightMeters / 2.0),
                        new Translation3d(-lengthMeters / 2.0, widthMeters / 2.0, -heightMeters / 2.0),
                        new Translation3d(-lengthMeters / 2.0, -widthMeters / 2.0, -heightMeters / 2.0)));
    }

    /**
     * Creates a spherical target model which has similar dimensions regardless of its rotation. This
     * model has four vertices:
     *
     * <ul>
     *   <li>Point 0: [0, -radius, 0]
     *   <li>Point 1: [0, 0, -radius]
     *   <li>Point 2: [0, radius, 0]
     *   <li>Point 3: [0, 0, radius]
     * </ul>
     *
     * <i>Q: Why these vertices?</i> A: This target should be oriented to the camera every frame, much
     * like a sprite/decal, and these vertices represent the ellipse vertices (maxima). These vertices
     * are used for drawing the image of this sphere, but do not match the corners that will be
     * published by photonvision.
     */
    public TargetModel(double diameterMeters) {
        double radius = diameterMeters / 2.0;
        this.vertices =
                List.of(
                        new Translation3d(0, -radius, 0),
                        new Translation3d(0, 0, -radius),
                        new Translation3d(0, radius, 0),
                        new Translation3d(0, 0, radius));
        this.isPlanar = false;
        this.isSpherical = true;
    }

    /**
     * Creates a target model from arbitrary 3d vertices. Automatically determines if the given
     * vertices are planar(x == 0). More than 2 vertices must be given. If this is a planar model, the
     * vertices should define a non-intersecting contour.
     *
     * @param vertices Translations representing the vertices of this target model relative to its
     *     pose.
     */
    public TargetModel(List<Translation3d> vertices) {
        this.isSpherical = false;
        if (vertices == null || vertices.size() <= 2) {
            vertices = new ArrayList<>();
            this.isPlanar = false;
        } else {
            boolean cornersPlanar = true;
            for (Translation3d corner : vertices) {
                if (corner.getX() != 0) cornersPlanar = false;
            }
            this.isPlanar = cornersPlanar;
        }
        this.vertices = vertices;
    }

    /**
     * This target's vertices offset from its field pose.
     *
     * <p>Note: If this target is spherical, use {@link #getOrientedPose(Translation3d,
     * Translation3d)} with this method.
     */
    public List<Translation3d> getFieldVertices(Pose3d targetPose) {
        var basisChange = new RotTrlTransform3d(targetPose.getRotation(), targetPose.getTranslation());
        return vertices.stream().map(basisChange::apply).collect(Collectors.toList());
    }

    /**
     * Returns a Pose3d with the given target translation oriented (with its relative x-axis aligned)
     * to the camera translation. This is used for spherical targets which should not have their
     * projection change regardless of their own rotation.
     *
     * @param tgtTrl This target's translation
     * @param cameraTrl Camera's translation
     * @return This target's pose oriented to the camera
     */
    public static Pose3d getOrientedPose(Translation3d tgtTrl, Translation3d cameraTrl) {
        var relCam = cameraTrl.minus(tgtTrl);
        var orientToCam =
                new Rotation3d(
                        0,
                        new Rotation2d(Math.hypot(relCam.getX(), relCam.getY()), -relCam.getZ()).getRadians(),
                        new Rotation2d(relCam.getX(), relCam.getY()).getRadians());
        return new Pose3d(tgtTrl, orientToCam);
    }

    @Override
    public boolean equals(Object obj) {
        return this == obj
                && obj instanceof TargetModel o
                && vertices.equals(o.vertices)
                && isPlanar == o.isPlanar
                && isSpherical == o.isSpherical;
    }
}
